File: src/matlab/weightsP.m
% SPDX-License-Identifier: GPL-3.0-only
% 
% Copyright 2008-2024 San Diego State University (SDSU) and Contributors 
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, version 3.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% LICENSE file or on the web GNU General Public License 
% <https://www.gnu.org/licenses/> for more details.
%
% ------------------------------------------------------------------------

function P = weightsP(k, m, dx)
% Returns the m+1 weights of P
%
% Parameters:
%                k : Order of accuracy
%                m : Number of cells
%               dx : Step size

    G = grad(k, m, dx);
    
    b = [-1; zeros(m, 1); 1]; % RHS
    
    P = G'\b;
end

File: src/cpp/weightsP.cpp
/*
* SPDX-License-Identifier: GPL-3.0-or-later
* © 2008-2024 San Diego State University Research Foundation (SDSURF).
* See LICENSE file or https://www.gnu.org/licenses/gpl-3.0.html for details. 
*/

/*
 * @file weights.cpp
 * 
 * @brief Mimetic Differences Q Weights
 * 
 * @date 2025/10/14
 * 
 */

#include "utils.h"
#include "divergence.h"
#include "gradient.h"
#include "weightsP.h"


WeightsP::WeightsP(u16 k, u32 m, Real dx)
{
  Gradient G(k, m, dx);

  vec b(m+2);
  b.at(0) = -1.0;
  b.at(m+1) = 1.0;

  sp_mat Gtranspose = G.t();
  *this = Utils::spsolve_eigenQR(Gtranspose,b);

}